feat: Added GPU node pool support for GKE, better logging, error and exception handling

mrburke00 opened the pull request:
- Automatic handling of GPU resources
     - Added node selectors and tolerations for when job requests GPU
     - Ensured GPU resource requests and limits are set appropriately in the pod specification 
     
- Better logging 
     - Logged complete pod specification before pod creation for easier debugging 

- Better exception handling 
    - Added exception handling around pod creation to catch and log API exceptions as well as job status checks 

- Resource requests and limit adjustments 
    -  Set both resource requests and limits for CPU, memory, and ephemeral storage to ensure correct scheduling    
    - Adjusted CPU requests to account for the `cpu_scalar`
    
- Resource definition now becomes 
```
resources:
    gpu=1,
    machine_type="g2-standard-8",
    mem_mb=8000
    threads=8   
```

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Introduced GPU resource requests for job specifications with manufacturer validation.
	- Added support for persistent volume mounts in job containers.
	- New documentation section on GPU scheduling capabilities for the Snakemake Kubernetes executor.

- **Improvements**
	- Enhanced logging for job execution and status checks, providing clearer error messages.
	- Improved error handling for unsupported GPU manufacturers and Kubernetes API calls.
	- Enhanced robustness of job execution and error reporting.
	- Detailed instructions for utilizing GPU resources in the documentation, including prerequisites for GKE setup.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

mrburke00 commented on snakemake_executor_plugin_kubernetes/__init__.py:
```diff
@@ -178,28 +175,52 @@ def run_job(self, job: JobExecutorInterface):
+        if 'gpu' in job.resources:
+            node_selector["accelerator"] = "nvidia"
+            self.logger.debug(f"GPU requested; updated node selector: {node_selector}")
+
```
Fixing

johanneskoester commented on snakemake_executor_plugin_kubernetes/__init__.py:
```diff
@@ -223,55 +244,64 @@ def run_job(self, job: JobExecutorInterface):
+        # Request GPU resources if specified
+        if "gpu" in job.resources.keys():
+            gpu_count = str(job.resources["gpu"])
+            container.resources.requests["nvidia.com/gpu"] = gpu_count
+            container.resources.limits["nvidia.com/gpu"] = gpu_count
+            self.logger.debug(f"Requested GPU resources: {gpu_count}")
```
What about non-nvidia GPUs? Maybe infer this from a `gpu_model: str` resource, which could allow types like `nvidia` or more specifically `tesla` etc.?

mrburke00 replied:
Well I specify the machine type already using the `machine_type="g2-standard-8`. But also I believe with GKE the node pool inherently specifies `"nvidia.com/gpu"`, without it the node scaling doesn't get triggered.

johanneskoester replied:
ok, but this plugin is meant to be generic, so we should try to ensure that added functionality works with any k8s cluster

mrburke00 replied:
Ive tested this method with every available GPU type offered via GKE. Its agnostic to GPU type and only looks for machine type. 

See the below K8 pod description.  Im selecting for machine type and not accelerator type. I can't think of any circumstance where you would select for an accelerator. You should be doing that via your terraform script or however else you build the cluster outside of snakemake.

```
autoscaling:
  enabled: true
  locationPolicy: BALANCED
  maxNodeCount: 5
config:
  accelerators:
  - acceleratorCount: '1'
    acceleratorType: nvidia-tesla-t4
    gpuDriverInstallationConfig:
      gpuDriverVersion: LATEST
  labels:
    accelerator: nvidia
    role: kubernetes
  loggingConfig: {}
  machineType: n1-standard-16
```

Its fine if you don't want to include this, but at least change the documentation to accurately reflect the fact that there is no GPU support offered via this plugin. I guarantee Im not the only one who put hundreds of hours of development time into a pipeline just to discover that there is no GPU support. Also BTW google life sciences is depreciated so there is no GPU support in general.

johanneskoester replied:
Sorry, I did not mean to put this down in any way. I just try to keep the k8s plugin independent of GKE. And I thought that this nvidia.com/gpu is very GKE specific and that we maybe would be able to offer a more generic implementation. As you are the expert here and not me, I might get it completely wrong though. Maybe we could schedule a short video meeting to discuss the details and help me understanding the problem? Just send me an email to johannes.koester@uni-due.de.

johanneskoester replied:
I have thought further about this, maybe I understand a bit better now. First, there seems to be an AMD plugin for k8s that handles gpu resources similarly: https://github.com/ROCm/k8s-device-plugin, also see https://github.com/ROCm/k8s-device-plugin/blob/master/example/pod/alexnet-gpu.yaml.
Second, I agree that the gpu model is maybe usually something that is controlled by the machine type. Also, it does not seem necessary to control the model itself via the workflow. What might make sense though is to control the manufacturer via the workflow, e.g. if you have a rule that uses cuda, you will want to annotate in your workflow that it requires an nvidia gpu.

So, maybe the following would work:
We suggest two standard resources `gpu=...` and `gpu_manufacturer=...`, where the latter is optional and might usually be just set via --default-resources gpu_manufacturer=nvidia or gpu_manufacturer=amd, depending on what kind of k8s cluster one has.
Then, in this plugin, we use `nvidia.com/gpu` in case the job has a resource `gpu_manufacturer=nvidia` and `amd.com/gpu` in the amd case.

Does that make sense, or do I miss something.

mrburke00 replied:
Yeah I can shoot you an email.

And yes that makes sense to me. Although it was only 6 months ago that a toolkit was established to allow AMD GPUs to execute CUDA code, which is what every scientific/ML tool uses for GPU acceleration. So the use cases for specifying gpu manufacturer might be pretty rare. It doesnt even appear that GKE offers AMD gpus.  

The other way to look at this is I'd imagine anyone getting this far is going to have a fairly advanced understanding of K8 and Terraform. So we could just establish enough resource specifications to construct the node selector themselves. I believe the current behavior is the scheduler just chooses the node with the minimum required computational resources, which is fine for most scenarios. But you could imagine a scenario when you want to choose a very specific node in a large cluster regardless of whether it has a gpu or not.

johanneskoester reviewed (changes requested):
Fantastic!

mrburke00 commented:
@johanneskoester In that newest commit I added a `manufacturer`resource type with support for nvidia and amd. An example for how the resource definition would now look is something like 
```
resources:
        gpu=1,
        manufacturer="nvidia",
        machine_type="n1-standard-16",
        mem_mb=32000,    
        threads=12
``` 
Tested with and without an NVIDIA GPU and the scaling works perfectly on GKE

johanneskoester commented on snakemake_executor_plugin_kubernetes/__init__.py:
```diff
@@ -178,28 +177,104 @@ def run_job(self, job: JobExecutorInterface):
+        # Add toleration for GPU nodes if GPU is requested
+        if "gpu" in resources_dict:
+            # Manufacturer logic
+            manufacturer = resources_dict.get("manufacturer", None)
+            if not manufacturer:
+                raise WorkflowError(
+                    "GPU requested but no manufacturer set. "
+                    "Use manufacturer='nvidia' or 'amd'."
```
```suggestion
            manufacturer = resources_dict.get("gpu_manufacturer", None)
            if not manufacturer:
                raise WorkflowError(
                    "GPU requested but no manufacturer set. "
                    "Use gpu_manufacturer='nvidia' or 'amd'."
```

johanneskoester replied:
Just to make this clear and distinguish it from other potential resources, I think it should be called gpu_manufacturer.

johanneskoester commented on snakemake_executor_plugin_kubernetes/__init__.py:
```diff
@@ -178,28 +177,104 @@ def run_job(self, job: JobExecutorInterface):
 
+        # Add node selector for GPU if requested
+        if "gpu" in resources_dict:
+            node_selector["accelerator"] = "nvidia"
```
Shouldn't this be changed now as well. since we now also support amd?

johanneskoester replied:
And how does it relate to this comment by coderabbit (might be nonsense of course)? https://github.com/snakemake/snakemake-executor-plugin-kubernetes/pull/31#discussion_r1934223200

mrburke00 replied:
Ah this was just a leftover debug statement I forgot. I will remove and push a new commit.

johanneskoester commented on snakemake_executor_plugin_kubernetes/__init__.py:
```diff
@@ -178,28 +177,99 @@ def run_job(self, job: JobExecutorInterface):
         )
+
+        # Add toleration for GPU nodes if GPU is requested
+        if "gpu" in resources_dict:
```
The resources considered here should be mentioned together with their effect for the resulting pods under docs/further.md.

mrburke00 replied:
I only see docs/intro.md do you want me to go ahead and create that file and adds the documentation?

johanneskoester replied:
Yes, that would be nice!

johanneskoester reviewed (changes requested):
Can you fix the formatting with black and the linting with flake8? Thanks! And some final things below.

mrburke00 commented on snakemake_executor_plugin_kubernetes/__init__.py:
```diff
@@ -223,55 +293,76 @@ def run_job(self, job: JobExecutorInterface):
+        if "disk_mb" in resources_dict:
+            disk_mb = int(resources_dict.get("disk_mb", 1024))
             container.resources.requests["ephemeral-storage"] = f"{disk_mb}M"
+            container.resources.limits["ephemeral-storage"] = f"{disk_mb}M"
```
I will remove them for now. I was mainly using them to catch if and when scaling were to be triggered. But you are right not absolutely necessary. Perhaps in a future iteration it would be worth it to allow users to configure that parameter

mrburke00 commented:
Removed limits and fixed some of the linting. I don't believe I have access to retrigger the linting checks so Ill wait to check if I fixed all of them.

mrburke00 commented:
Added `docs/further.md` with relevant documentation. Looks like the formatting checks are good to go as well. @johanneskoester let me know if you need anything else from me.

johanneskoester commented:
Very nice, thanks a lot!
